Sources/URLHandler.swift

```diff
@@ -0,0 +1,203 @@
+                    UIApplication.shared.open(callback, options: [:], completionHandler: nil)
+                } else {
+                    UIApplication.shared.openURL(callback)
+                }
+            }
+            return true
+        } else if action == "download", let movieURL = movieURL {
+            downloadMovie(from:movieURL, fileNameOfMedia:fileName)
+            return true
+        }
+        return false
+    }
+}
+
+class VLCCallbackURLHandler: NSObject, VLCURLHandler {
+
+    @objc func canHandleOpen(url: URL, options: [UIApplicationOpenURLOptionsKey: AnyObject]) -> Bool {
+        return url.scheme == "vlc"
+    }
+
+    // Safari fixes URLs like "vlc://http://example.org" to "vlc://http//example.org"
+    func transformVLCURL(_ url: URL) -> URL {
+        var parsedString = url.absoluteString.replacingOccurrences(of: "vlc://", with: "")
+        if let location = parsedString.range(of: "//"), parsedString[parsedString.index(location.lowerBound, offsetBy: -1)] != ":" {
+            parsedString = "\(parsedString[parsedString.startIndex..<location.lowerBound])://\(parsedString[location.upperBound...])"
+        } else if !parsedString.hasPrefix("http://") && !parsedString.hasPrefix("https://") && !parsedString.hasPrefix("ftp://") {
+            parsedString = "http://\(parsedString)"
+        }
+        return URL(string: parsedString)!
+    }
+
+    func performOpen(url: URL, options: [UIApplicationOpenURLOptionsKey: AnyObject]) -> Bool {
+
+        let transformedURL = transformVLCURL(url)
+        let scheme = transformedURL.scheme
+        if scheme == "http" || scheme == "https" || scheme == "ftp" {
+            let alert = UIAlertController(title: NSLocalizedString("OPEN_STREAM_OR_DOWNLOAD", comment:""), message: url.absoluteString, preferredStyle: .alert)
+            let downloadAction = UIAlertAction(title: NSLocalizedString("BUTTON_DOWNLOAD", comment:""), style: .default) { _ in
+                downloadMovie(from:transformedURL, fileNameOfMedia:nil)
+            }
+            alert.addAction(downloadAction)
+            let playAction = UIAlertAction(title: NSLocalizedString("PLAY_BUTTON", comment:""), style: .default) { _ in
+                play(url: transformedURL, completion: nil)
+            }
+            alert.addAction(playAction)
+            alert.show(UIApplication.shared.keyWindow!.rootViewController!, sender: nil)
+        } else {
+            play(url: transformedURL, completion: nil)
+        }
+        return true
+    }
+}
+
+class ElseCallbackURLHandler: NSObject, VLCURLHandler {
+    @objc func canHandleOpen(url: URL, options: [UIApplicationOpenURLOptionsKey: AnyObject]) -> Bool {
+        return true
+    }
+
+    func performOpen(url: URL, options: [UIApplicationOpenURLOptionsKey: AnyObject]) -> Bool {
+        play(url: url, completion: nil)
+        return true
+    }
+}
+
+// TODO: This code should probably not live here
+func play(url: URL, completion: ((Bool) -> Void)?) {
```

Comment:
I realise that there is a `TODO` for it  but I really think we should encapsulated them even temporarily.
Form what I understand the two methods are only used by classes that conforms to `VLCURLHandler`.
What would you think adding them inside an extension of  `VLCURLHandler` like so: 


```extension VLCURLHandler {
    // TODO: This code should probably not live here
    func play(url: URL, completion: ((Bool) -> Void)?) {
      ...
    }

    func downloadMovie(from url: URL, fileNameOfMedia fileName: String?) {
        ...
    }
}
```

On the other hand, the actual call of the methods would have to be changed a tiny bit.